File: hunter/hunter/backends/omp_scavenge/capacity.py
"""Budget policy -- two linear ramps.

7-day ramp:  allowed = elapsed_fraction_of_7d_window.
             Spreads spending evenly across the week.

5-hour ramp: allowed = max(0, (elapsed - HEADROOM) / (5h - HEADROOM)).
             Zero for the first HEADROOM duration (human headroom), then 0→1
             over the remaining time.  Anything unspent at reset is wasted capacity.

Both horizons reduce to one comparison: allowed = ramp(...) > effective_used.
"exhausted" is folded into effective_used (clamped to exactly 1.0, see
_effective_used) rather than special-cased -- since both ramps are
themselves capped at 1.0 (reached only exactly at resets_at), an exhausted
window denies for its whole remaining duration for free, with no
ramp-catchup-before-reset risk.

A window's last known reading is never discarded for being old -- usage
only increases within a window, so it remains a valid floor. Combined
with unaccounted-token tracking (computed by the facade from the
SpendLedger, tracked separately per window), and ramps that keep
climbing toward 1.0 against live wall-clock time regardless of probe
staleness, denial resolves itself once the ramp naturally catches up --
no bypass needed.

No active 5h window (truly nothing ever probed, e.g. a fresh install) →
allow (opens one), gated only by the 7d ramp. A window whose own
resets_at has passed is NOT "no active window" -- read_windows() rolls
anthropic:5h/anthropic:7d forward into their current cycle instead of
dropping them (a genuinely abandoned per-model-class dimension still
gets dropped; see read_windows). Missing data entirely → deny.
"""

from __future__ import annotations

import sqlite3
import time
from dataclasses import dataclass, field
from pathlib import Path

from hunter.types import Config

# ---------------------------------------------------------------------------
# Locally-defined constants and types (decoupled from hunter.types)
# ---------------------------------------------------------------------------

OMP_AGENT_DB = Path.home() / ".omp/agent/agent.db"


@dataclass
class WindowState:
    limit_id: str
    used_fraction: float | None
    status: str | None  # ok | exhausted | ...
    resets_at: int | None  # epoch ms
    recorded_at: int  # epoch ms -- when omp probed it
    age_s: float = field(default=0.0)

    @property
    def stale(self) -> bool:
        return self.age_s > 1800



# =============================================================================
# Configuration: To adjust when hunter can start using the 5h window,
#                change HEADROOM_MS below (e.g., 15min, 1h, 2h).
#                Everything else auto-computes from it.
# =============================================================================
HEADROOM_MS = 30 * 60 * 1000  # Human headroom before harvest window opens

_WEEK_MS = 7 * 24 * 3600 * 1000
_5H_MS = 5 * 3600 * 1000
_RAMP_MS = _5H_MS - HEADROOM_MS  # Harvest window duration (4.5h at 30min headroom)


def read_windows() -> dict[str, WindowState]:
    """Latest usage_history row per anthropic:* limit. Missing DB -> {}."""
    if not OMP_AGENT_DB.exists():
        return {}
    try:
        db = sqlite3.connect(f"file:{OMP_AGENT_DB}?mode=ro", uri=True)
        db.row_factory = sqlite3.Row
        rows = db.execute(
            "SELECT limit_id, used_fraction, status, resets_at,"
            " MAX(recorded_at) AS recorded_at"
            " FROM usage_history WHERE limit_id LIKE 'anthropic:%'"
            " GROUP BY limit_id"
        ).fetchall()
        db.close()
    except sqlite3.Error:
        return {}
    now = time.time() * 1000
    out: dict[str, WindowState] = {}
    for r in rows:
        resets_at = r["resets_at"]
        used_fraction = r["used_fraction"]
        status = r["status"]
        recorded_at = r["recorded_at"]
        if resets_at and resets_at <= now:
            # This window's own cycle has ended -- but that does NOT mean
            # "no active window": 5h/7d windows are back-to-back, so a new
            # cycle is definitely running RIGHT NOW, hunter just hasn't
            # been probed for it yet. Dropping the row entirely (the old
            # behavior) was a real production incident: decide()'s
            # unaccounted_tokens reservation exists exactly to survive a
            # stale-but-still-current probe, but it only ever gets
            # consulted when a WindowState is present to attach it to --
            # once this function dropped the row, decide() saw no
            # anthropic:5h data at all and fell through to its "no active
            # window -> allow" rule, completely bypassing
            # unaccounted_tokens regardless of how large it had grown. A
            # cascade of modernization jobs (which routinely cost 2-5x
            # their nominal cap on a cold-cache first call, see
            # runner.py) burned ~2.9M tokens across a freshly-rolled-over
            # 5h window with zero denials, because nothing was left to
            # deny against.
            #
            # Fix: roll the window forward ourselves instead of dropping
            # it. A brand-new window legitimately starts at 0% (Anthropic's
            # own probe will confirm that once it lands) -- what's NOT
            # legitimate is treating "0% probed" as "0% spent". recorded_at
            # is set to the boundary the new cycle actually started at (not
            # `now`), so _unaccounted_tokens's "finished since the last
            # probe" query correctly sums every job hunter has run since
            # the rollover, not just since this function last happened to
            # run.
            # Scoped to the exact account-wide dimensions decide() actually
            # gates real work on -- NOT per-model-class variants like
            # anthropic:7d:fable, which really can go genuinely abandoned
            # (config stopped routing there) and must stay dropped, both to
            # avoid decide() noise and to keep the Status page from showing
            # a synthesized "fresh" window for a dimension nothing uses.
            period = {"anthropic:5h": _5H_MS, "anthropic:7d": _WEEK_MS}.get(r["limit_id"])
            if period is None:
                continue  # per-model-class or unrecognized -- keep the old, safe drop
            new_resets_at = resets_at
            while new_resets_at <= now:
                recorded_at = new_resets_at
                new_resets_at += period
            resets_at = new_resets_at
            used_fraction = 0.0
            status = "ok"
        out[r["limit_id"]] = WindowState(
            limit_id=r["limit_id"],
            used_fraction=used_fraction,
            status=status,
            resets_at=resets_at,
            recorded_at=recorded_at,
            age_s=(now - recorded_at) / 1000,
        )
    return out


def ramp_7d(resets_at: int | None, now_ms: float) -> float:
    """Fraction of the 7-day linear ramp elapsed so far (0..1) -- this is
    the currently-allowed spend ceiling for a :7d window, the same
    quantity decide() compares used_fraction against. Unknown or already-
    expired resets_at -> 1.0 (assume end-of-window; matches decide()'s
    existing permissive fallback for missing data, and read_windows()
    already drops genuinely-expired rows before this could be called
    with stale data)."""
    if not resets_at or resets_at <= now_ms:
        return 1.0
    started_ms = resets_at - _WEEK_MS
    return min((now_ms - started_ms) / _WEEK_MS, 1.0)


def ramp_5h(resets_at: int | None, now_ms: float) -> float | None:
    """Fraction of the 5h harvest ramp elapsed so far (0..1): 0 for the
    first HEADROOM_MS, then linear to 1.0 at reset. None means there is
    no active window to compute a ramp against (decide() then treats the
    window as an opener -- always-allow)."""
    if not resets_at or resets_at <= now_ms:
        return None
    elapsed_ms = _5H_MS - (resets_at - now_ms)
    return max(0.0, (elapsed_ms - HEADROOM_MS) / _RAMP_MS)


def retry_at_7d(resets_at: int | None, effective_used: float) -> float | None:
    """Epoch ms when the 7d ramp would reach effective_used (assuming no
    further spend) -- the earliest a 7d-ramp denial at this used level
    could resolve. None if resets_at is unknown: no informed estimate is
    possible, callers should fall back to a generic backoff rather than
    a value that looks precise but isn't."""
    if not resets_at:
        return None
    started_ms = resets_at - _WEEK_MS
    return started_ms + effective_used * _WEEK_MS


def retry_at_5h(resets_at: int | None, effective_used: float) -> float | None:
    """Epoch ms when the 5h harvest ramp would reach effective_used
    (assuming no further spend). None if resets_at is unknown."""
    if not resets_at:
        return None
    window_start_ms = resets_at - _5H_MS
    return window_start_ms + HEADROOM_MS + effective_used * _RAMP_MS


def _effective_used(w: WindowState, inflight_reservation: float) -> float:
    """used_fraction, adjusted for spend the probe can't see yet.

    "exhausted" is clamped to exactly 1.0 rather than trusting the raw
    reported fraction: it's Anthropic's hard-stop signal, not a pacing
    number, and its literal value isn't reliable across limit types
    (observed as high as 1.57 for a non-Anthropic limit in this same
    usage_history table). Trusting it verbatim could let a fraction just
    under 1.0 be second-guessed by ramp catch-up before the window is
    actually over, or push a ramp-derived retry_at past the real reset.
    """
    used = 1.0 if w.status == "exhausted" else w.used_fraction
    return used + inflight_reservation

File: hunter-rs/src/backends/omp_scavenge/capacity.rs
//! Window reading + pure ramp math (capacity.py, BACKEND-CONTRACT.md §2.1).
//! Pure fns stay parameterized on `now_ms` (deterministic tests);
//! `read_windows` takes the agent.db path explicitly (tests point it at a
//! fixture).

use super::provider::LlmProvider;
use sqlx::Row;
use std::collections::BTreeMap;

use std::path::Path;

/// THE tunable: human headroom at 5h-window start (`capacity.HEADROOM_MS`).
pub const HEADROOM_MS: i64 = 30 * 60 * 1000;
pub const WEEK_MS: i64 = 604_800_000;
pub const FIVE_H_MS: i64 = 18_000_000;
/// 5h ramp span after headroom: 4.5 h.
pub const RAMP_MS: i64 = FIVE_H_MS - HEADROOM_MS;

/// One provider window as read from omp's usage mirror.
#[derive(Debug, Clone)]
pub struct WindowState {
    pub limit_id: String,
    pub used_fraction: Option<f64>,
    pub status: Option<String>,
    pub resets_at: Option<i64>,
    /// When omp probed (epoch ms) — or, for rolled-forward expired
    /// cycles, the current cycle's start boundary.
    pub recorded_at: i64,
    pub age_s: f64,
}

/// Default agent.db location (`capacity.OMP_AGENT_DB`); callers may override.
pub fn default_agent_db() -> std::path::PathBuf {
    let home = std::env::var_os("HOME")
        .map(std::path::PathBuf::from)
        .unwrap_or_default();
    home.join(".omp/agent/agent.db")
}

/// Read newest `usage_history` row per anthropic:% `limit_id`; roll expired
/// account-wide cycles forward (used=0, status=ok, `recorded_at=cycle`
/// start), DROP expired per-model-class rows. Missing file or ANY sqlite
/// error -> empty map. `BTreeMap`: deny-reason precedence needs ascending
/// `limit_id` order.
///
/// An empty map is indistinguishable from "no window data" downstream:
/// `decide` denies until fresh and `keep_fresh` probes forever. A genuine
/// read failure therefore looks exactly like an idle daemon unless it is
/// logged here. A missing file or a row-less table is not a failure and
/// stays silent.
pub fn read_windows(agent_db: &Path, now_ms: i64) -> BTreeMap<String, WindowState> {
    read_windows_for(agent_db, LlmProvider::Anthropic, now_ms)
}

pub fn read_windows_for(
    agent_db: &Path,
    provider: LlmProvider,
    now_ms: i64,
) -> BTreeMap<String, WindowState> {
    let Ok(handle) = tokio::runtime::Handle::try_current() else {
        tracing::error!(
            agent_db = %agent_db.display(),
            "read_windows called outside a Tokio runtime — usage windows unavailable"
        );
        return BTreeMap::new();
    };
    match handle.block_on(read_windows_async(agent_db, provider, now_ms)) {
        Ok(windows) => windows,
        Err(err) => {
            tracing::error!(
                agent_db = %agent_db.display(),
                error = %err,
                "reading omp usage windows failed — usage windows unavailable"
            );
            BTreeMap::new()
        }
    }
}

/// Async inner: opens `agent_db` read-only, queries newest row per
/// `limit_id` via window function, rolls forward expired account-wide
/// cycles, drops expired per-model-class rows.
async fn read_windows_async(
    agent_db: &Path,
    provider: LlmProvider,
    now_ms: i64,
) -> Result<BTreeMap<String, WindowState>, Box<dyn std::error::Error + Send + Sync>> {
    use sqlx::sqlite::SqliteConnectOptions;
    if !agent_db.exists() {
        return Ok(BTreeMap::new());
    }

    let opts = SqliteConnectOptions::new()
        .filename(agent_db)
        .read_only(true);
    let pool = sqlx::SqlitePool::connect_with(opts).await?;

    let limits = provider.windows();
    let rows = sqlx::query(
        "SELECT limit_id, used_fraction, status, resets_at, recorded_at \
         FROM ( \
           SELECT limit_id, used_fraction, status, resets_at, recorded_at, \
                  ROW_NUMBER() OVER (PARTITION BY limit_id ORDER BY recorded_at DESC) AS rn \
           FROM usage_history \
           WHERE provider = ? AND limit_id IN (?, ?) \
         ) WHERE rn = 1",
    )
    .bind(provider.name())
    .bind(limits.short.limit_id)
    .bind(limits.long.limit_id)
    .fetch_all(&pool)
    .await?;

    pool.close().await;

    let mut result = BTreeMap::new();

    for row in rows {
        // agent.db belongs to omp, so its schema and values are foreign
        // input: `Row::get` would panic on a NULL or a wrongly-typed cell
        // and unwind the caller's `spawn_blocking`, turning one bad cell
        // into a JoinError on every `/api/summary` poll and every worker
        // pre-snapshot. A decode failure abandons the WHOLE snapshot
        // rather than the offending row, because a partial map is unsafe
        // in a way an empty one is not: `decide_inner` only refuses when
        // the map is empty, so dropping just an undecodable anthropic:5h
        // row would let spending be granted against a window nobody can
        // see, while the empty map is the designed "capacity unknown"
        // state (deny until fresh, probe, "no window data" in the UI).
        let limit_id: String = row.try_get("limit_id")?;
        let used_fraction: Option<f64> = row.try_get("used_fraction")?;
        let status: Option<String> = row.try_get("status")?;
        let resets_at: Option<i64> = row.try_get("resets_at")?;
        let recorded_at: i64 = row.try_get("recorded_at")?;

        match resets_at {
            // Truthy resets_at (> 0) that has expired (<= now_ms).
            // Some(0) is falsy like Python's `not resets_at`.
            Some(r) if r > 0 && r <= now_ms => {
                let period = if limit_id == limits.short.limit_id {
                    limits.short.period_ms
                } else if limit_id == limits.long.limit_id {
                    limits.long.period_ms
                } else {
                    continue;
                };
                // Roll forward: advance resets_at by period until > now,
                // recording each step as the current cycle's start.
                let mut new_resets = r;
                let mut new_recorded = recorded_at;
                while new_resets <= now_ms {
                    new_recorded = new_resets;
                    new_resets += period;
                }
                result.insert(
                    limit_id.clone(),
                    WindowState {
                        limit_id,
                        used_fraction: Some(0.0),
                        status: Some("ok".to_owned()),
                        resets_at: Some(new_resets),
                        recorded_at: new_recorded,
                        age_s: (now_ms - new_recorded) as f64 / 1000.0,
                    },
                );
            }
            _ => {
                // Not expired, NULL resets_at, or Some(0) → keep verbatim.
                result.insert(
                    limit_id.clone(),
                    WindowState {
                        limit_id,
                        used_fraction,
                        status,
                        resets_at,
                        recorded_at,
                        age_s: (now_ms - recorded_at) as f64 / 1000.0,
                    },
                );
            }
        }
    }

    Ok(result)
}

/// not `resets_at` or expired -> 1.0; else min(elapsed/week, 1.0).
pub fn ramp_7d(resets_at: Option<i64>, now_ms: i64) -> f64 {
    match resets_at {
        None | Some(0) => 1.0,
        // An expired window needs no arm of its own: elapsed is then at
        // least a full week, so the `.min(1.0)` below already returns 1.0.
        // A guard here would be unfalsifiable — no input distinguishes it.
        Some(r) => {
            let elapsed = (now_ms - (r - WEEK_MS)) as f64;
            (elapsed / WEEK_MS as f64).min(1.0)
        }
    }
}

/// not `resets_at` or expired -> None (no active window: opener allowed);
/// else max(0, (elapsed - HEADROOM) / RAMP).
pub fn ramp_5h(resets_at: Option<i64>, now_ms: i64) -> Option<f64> {
    match resets_at {
        None | Some(0) => None,
        Some(r) if r <= now_ms => None, // expired
        Some(r) => {
            let elapsed = (FIVE_H_MS - (r - now_ms)) as f64;
            Some(((elapsed - HEADROOM_MS as f64) / RAMP_MS as f64).max(0.0))
        }
    }
}

/// Exact inverse of `ramp_7d`; None when `resets_at` is None/0.
pub fn retry_at_7d(resets_at: Option<i64>, effective_used: f64) -> Option<f64> {
    match resets_at {
        None | Some(0) => None,
        Some(r) => {
            let raw = (r - WEEK_MS) as f64 + effective_used * WEEK_MS as f64;
            Some(raw.min(r as f64))
        }
    }
}

pub fn retry_at_5h(resets_at: Option<i64>, effective_used: f64) -> Option<f64> {
    match resets_at {
        None | Some(0) => None,
        Some(r) => {
            let raw = (r - FIVE_H_MS) as f64 + HEADROOM_MS as f64 + effective_used * RAMP_MS as f64;
            Some(raw.min(r as f64))
        }
    }
}

/// status == "exhausted" clamps to exactly 1.0 (hard-stop signal; raw
/// value unreliable); else `used_fraction` (call sites guarantee Some) +
/// inflight reservation.
pub fn effective_used(w: &WindowState, inflight_reservation: f64) -> f64 {
    let used = if w.status.as_deref() == Some("exhausted") {
        1.0
    } else {
        w.used_fraction.unwrap_or(0.0)
    };
    used + inflight_reservation
}
